worlds/wl4/__init__.py

```diff
@@ -0,0 +1,240 @@
+
+    def create_regions(self):
+        location_table = self.setup_locations()
+        create_regions(self, location_table)
+        set_access_rules(self)
+        connect_regions(self)
+
+        passages = ('Emerald', 'Ruby', 'Topaz', 'Sapphire')
+        for passage in passages:
+            location = self.multiworld.get_region(f'{passage} Passage Boss', self.player).locations[0]
+            location.place_locked_item(self.create_item(f'{passage} Passage Clear'))
+            location.show_in_spoiler = False
+
+        if self.options.goal.needs_diva():
+            goal = 'Golden Diva'
+        if self.options.goal.is_treasure_hunt():
+            goal = 'Sound Room - Emergency Exit'
+
+        goal = self.multiworld.get_location(goal, self.player)
+        goal.place_locked_item(self.create_item('Escape the Pyramid'))
+        goal.show_in_spoiler = False
+
+    def create_items(self):
+        difficulty = self.options.difficulty
+        treasure_hunt = self.options.goal.needs_treasure_hunt()
+
+        gem_pieces = 18 * 4
+        cds = 16
+        full_health_items = (9, 7, 6)[difficulty.value]
+        treasures = 12 * treasure_hunt
+        total_required_locations = gem_pieces + cds + full_health_items + treasures
+
+        itempool = []
+
+        required_jewels = self.options.required_jewels.value
+        pool_jewels = self.options.pool_jewels.value
+        for name, item in filter_items(type=ItemType.JEWEL):
+            force_non_progression = required_jewels == 0
+            if item.passage() == Passage.ENTRY:
+                copies = min(pool_jewels, 1)
+            elif item.passage() == Passage.GOLDEN:
+                copies = self.options.golden_jewels.value
+                if self.options.goal.is_treasure_hunt():
+                    force_non_progression = True
+            else:
+                copies = pool_jewels
+
+            for _ in range(copies):
+                itempool.append(self.create_item(name, force_non_progression))
+
+        for name in filter_item_names(type=ItemType.CD):
+            itempool.append(self.create_item(name))
+
+        for name in filter_item_names(type=ItemType.ABILITY):
+            itempool.append(self.create_item(name))
+            if name.startswith('Progressive'):
+                itempool.append(self.create_item(name))
+
+        # Remove full health items to make space for abilities
+        if required_jewels == 4:
+            if difficulty == 0:
+                full_health_items -= 8
+            else:
+                raise ValueError('Not enough locations to place abilities for '
+                                 f'{self.multiworld.player_name[self.player]}. '
+                                 'Set the "Required Jewels" setting to a lower '
+                                 'value and try again.')
+
+        for _ in range(full_health_items):
+            itempool.append(self.create_item('Full Health Item'))
+
+        if (treasure_hunt):
```

Comment:
These are few places with redundant parens, but this one struck me as especially unneeded
```suggestion
        if treasure_hunt:
```